First version of the docker configuration

guenoledc opened the pull request:
Hi @dev-adamant-im, I have pushed a PR.

Please let us know.
I will be trying to use this setup for my current development solution

guenoledc committed:
First version of the docker configuration

drewwells commented on docker-compose.yml:
```diff
@@ -0,0 +1,48 @@
+    volumes:
+      - ./create_table.sql:/docker-entrypoint-initdb.d/init.sql
+
+  postgrest:
```
```suggestion
  postgres:
```

adamant-al replied:
I think `postgrest` is correct, it means `postgREST`

guenoledc replied:
`db` is the postgre database and `postgrest` is the tool to expose tables as REST api

drewwells commented on docker-compose.yml:
```diff
@@ -0,0 +1,48 @@
+      POSTGRES_DB: app_db
+      POSTGRES_USER: app_user
+      POSTGRES_PASSWORD: postgres!secret
+  # Uncomment this if you want to persist the data.
+  # volumes:
+  #   - "./pgdata:/var/lib/postgresql/data"
```
Persistence should be the default, it's also important to nest it in a folder. postgresql sets ownership on the directory and it will be impossible to dockerignore without nesting the folder and ignoring the parent folder.
```suggestion
    volumes:
      - ./data/postgres:/var/lib/postgresql/data
```

drewwells commented on docker-compose.yml:
```diff
@@ -0,0 +1,48 @@
+services:
+
+  db:
+    image: postgres:10-alpine
```
Not sure what version we need, but 10 is pretty old. Switched to 12
```suggestion
    image: postgres:12
```

drewwells commented on docker-compose.yml:
```diff
@@ -0,0 +1,48 @@
+  # volumes:
+  #   - "./pgdata:/var/lib/postgresql/data"
+    volumes:
+      - ./create_table.sql:/docker-entrypoint-initdb.d/init.sql
```
Based on my reading, this only executes if the database has not yet been created ie. /data/postgres is empty. I think this is fine. Ideally, we could run migrations and what not from the application, but this should get the job done.

guenoledc replied:
Agreed, but that would require a bit more changes than the docker embedding and I feel it should be another issue and change

drewwells reviewed (changes requested):
Can you enable persistence by default? Restarting the container shouldn't drop the database and start from scratch.

adamant-al commented on README.md:
````diff
@@ -262,6 +262,23 @@ Zabbix API availability trigger examples:
 {API for Ethereum transactions:system.run["curl -s --connect-timeout 7 https://ethnode1.adamant.im/max_block | jq .[].max"].change()}<1
 ```
 
+# Dockerized and docker compose
+by Guénolé de Cadoudal (guenoledc@yahoo.fr)
+
+In the `docker-compose.yml` you find a configuration that show how this tool can be embeded in a docker configuration with the following processes
+- postgres db: to store the data
+- postgREST tool to expose the data as a REST api (see above comments)
+- GETH node in POA mode. Can be anothe node of course bu not tested
+- EthSync tool (this tool)
+
+The EthSync tool accept the following env variables
+- DB_NAME: postgres url of the db
+- ETH_URL: eth node url to reach the node. support websocket, http and ipc
+- START_BLOCK: the first block to synchronize from. default to 1
+- CONFIRMATIONS_BLOCK: the number of blocks to leave out of the synch from the end. Ie, last block is current `blockNumber - CONFIRMATIONS_BLOCK`. Default is 0.
+- PERIOD: Number of seconds between to synchronization. Default is 20 sec
+
+
````
```suggestion
# Dockerized and docker compose
by Guénolé de Cadoudal (guenoledc@yahoo.fr)

In the `docker-compose.yml` you find a configuration that show how this tool can be embedded in a docker configuration with the following processes:
- postgres db: to store the indexed data
- postgREST tool to expose the data as a REST api (see above comments)
- GETH node in POA mode. Can be Openethereum, or another node, but not tested
- EthSync tool (this tool)

The EthSync tool accepts the following env variables:
- DB_NAME: postgres url of the db
- ETH_URL: eth node url to reach the node. Supports websocket, http and ipc.
- START_BLOCK: the first block to synchronize from. Default is 1.
- CONFIRMATIONS_BLOCK: the number of blocks to leave out of the synch from the end. I.e., last block is current `blockNumber - CONFIRMATIONS_BLOCK`. Default is 0.
- PERIOD: Number of seconds between to synchronization. Default is 20 sec.

```

adamant-al commented on docker-compose.yml:
```diff
@@ -0,0 +1,48 @@
+    depends_on:
+      - db
+    
+
```
```suggestion
```

adamant-al commented on ethsync.py:
```diff
@@ -5,32 +5,53 @@
 import logging
 #from systemd.journal import JournalHandler
 
 # Get postgre database name
```
```suggestion
# Get env variables or set to default
```

adamant-al commented on ethsync.py:
```diff
@@ -5,32 +5,53 @@
+    print('Add eth url in env var ETH_URL')
+    exit(2)
 
 # Connect to geth node
 #web3 = Web3(Web3.IPCProvider("/home/geth/.ethereum/geth.ipc"))
 
 # Or connect to openethereum node
```
```suggestion
# Connect to Ethereum node
```

adamant-al commented on ethsync.py:
```diff
@@ -5,32 +5,53 @@
+if web3 == None:
+    web3 = Web3(Web3.IPCProvider(nodeUrl)) # "/home/parity/.local/share/openethereum/jsonrpc.ipc"
+
+
```
```suggestion

```

adamant-al commented on ethsync.py:
```diff
@@ -113,9 +136,9 @@ def insertion(blockid, tr):
     maxblockindb = cur.fetchone()[0]
     # On first start, we index transactions from a block number you indicate. 10000000 is a sample.
```
```suggestion
    # On first start, we index transactions from a block number you indicate
```

adamant-al reviewed (commented):
Looks good!

guenoledc committed:
Apply suggestions from code review

Co-authored-by: adamant-al <33592982+adamant-al@users.noreply.github.com>
Co-authored-by: Drew Wells <drew.wells00@gmail.com>

guenoledc committed:
Adjusting to reviewers comments and few changes

guenoledc commented:
In the last commit I have added your change requested.

Also, I have added 2 changes
- Enable launching the ethsync tool only when the geth node and db are ready (with health checks)
- forcing the conversion of the transaction hash to hex because else it was recorded as binary data and I need it as `0x....`

guenoledc commented:
I've created the image in dockerhub to avoid being linked to the source code @ https://hub.docker.com/r/guenoledc/eth-tx-storage

adamant-al committed:
Update contributors

adamant-al reviewed (commented):
May be you add a link `https://hub.docker.com/r/guenoledc/eth-tx-storage` somewhere in description? 
After you do some testing, let me know, we'll merge this PR.

adamant-al committed:
Update contributors

guenoledc commented:
> May be you add a link `https://hub.docker.com/r/guenoledc/eth-tx-storage` somewhere in description?
> After you do some testing, let me know, we'll merge this PR.

I have pushed the image to dockerhub but I feel it should be pushed from your repo after the merge is done so I don't have to maintain my repo.

adamant-al commented:
@guenoledc Hi! Have you tested the docker changes?
We are ready to merge.

guenoledc commented:
> @guenoledc Hi! Have you tested the docker changes?
> We are ready to merge.

@adamant-al yes I have used the created image in my project. It is not necessarily production ready but it works
